File: Java/historical_tolls.java
import java.io.File;
import java.io.PrintWriter;

/**
 * Created by Sung Kim on 3/2/2016.
 *
 * Usage: java historical_tolls <num xways> <max carid> <outfile>
 * NOTE: num xway is the number of xways, not the 0-indexed largest xway.  So, 3 files => java historical_tolls 3 <max carid> <outfile>
 */

public class historical_tolls {
    final static int NUM_DAYS_IN_HISTORY = 70;

    public static void main(String[] args) throws Exception {
        int maxXway = Integer.parseInt(args[0]);
        int maxCarId = Integer.parseInt(args[1]);
        File outfile = new File(args[2]);
        int i, day, toll, xway;
        maxCarId++; // we want to include the max carid in the result set
        PrintWriter writer = new PrintWriter(outfile);
        for (i = 0; i < maxCarId; i++) {
            for (day = 1; day < NUM_DAYS_IN_HISTORY; day++) {
                toll = (int) (Math.random() * 1000) % 90 + 10;
                xway = (int) (Math.random() * 1000) % maxXway;
                writer.println(i + "," + day + "," + xway + "," + toll);
            }
        }
        writer.close();
    }
}

File: C/historical_tolls.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Author: Sung Kim
//
// Description: This is a single-threaded C version.
// Toll generation for 1M cars with 200 XWays takes ~21 seconds.
// On the same HW the original perl version took ~45 seconds.
// On the same HW the Java version takes ~15 seconds.
// On the same HW the Java version, but using System.out instead of PrintWriter takes ~4 min.
// On the same HW the C version, using fprintf on a file instead of printf and redirection takes ~46 sec.



const int NUM_DAYS_IN_HISTORY = 70;

int main(int argc, char* argv[])
{
    if (argc != 4)
    {
        printf("Usage: ./historical_tolls <Max XWay> <Max CarId> <Output file>\n");
        printf("<Max XWay> is [0, max] and <Max CarId> is [0, max) \n");
        return 1;
    }

	int max_xway = atoi(argv[1]);
	int max_carid = atoi(argv[2]);
	char *outfile = argv[3];

	printf("Max XWay: %d\n", max_xway);
	printf("Max CarId: %d\n", max_carid);
	printf("Outfile: %s\n", outfile);

	int i, day, toll, xway;
	//FILE *writer;

	/*
	writer = fopen(outfile, "w");
	if (writer == NULL)
	{
	    fprintf(stderr, "Can't open file %s for writing. Exiting.\n", outfile);
	    return 1;
	}
	*/

	srand(time(NULL));

	max_xway++;  // We do this we when we % we get 0-max_xway xways.

	for (i = 0 ; i < max_carid ; i++)
	{
		for (day = 1 ; day < NUM_DAYS_IN_HISTORY ; day++)
		{
            clock_t time4rand = clock();
			toll = rand() % 90 + 10;  // The +1 gives us up to 99 and removes 0 tolls
			xway = rand() % max_xway;  // The original was ] (inclusive, so this one is too)
			printf("%d,%d,%d,%d\n",i,day,xway,toll);
			//fprintf(writer, "%d,%d,%d,%d\n",i,day,xway,toll);
			//printf("Time to create random toll and random xway and print it%f\n",
            //			((double) (clock() - time4rand)) / CLOCKS_PER_SEC);
		}
	}

	//fclose(writer);

	return 0;
}


